### DISCUSSION TITLE
How to send nested json data in server action?

### DISCUSSION THREAD

**Discussion by NiazMorshed2007** [vercel/next.js#66341] [Help]
### Summary

With server action it enables to send form data to the server, which sounds simple.
But I had a question of what if I want to send nested json data with the form?

First looking at the simple cases:

```jsx
export default function Page() {
  async function createInvoice(formData: FormData) {
    'use server'
 
    const rawFormData = {
      customerId: formData.get('customerId'),
      amount: formData.get('amount'),
      status: formData.get('status'),
    }
 
    // mutate data
    // revalidate cache
  }
 
  return <form action={createInvoice}>...</form>
}
```

But now let's say I want to pass a `address` field in the form action. However `address` is an object. It has own properties like this:

```
address {
  lat,
  lng,
  state,
  zipcode,
  ...etc
}
```

Now the question is, how do I send it to the server? I can assume I can `JSON.stringify(address)` the `address`.
And then get in the server action like so: `JSON.parse(formData.get("address"))`.

This works, but is there any better approaches than this?

Thanks. 🙏

**Comment by icyJoseph** [comment]
`stringify` works only for plain objects, if there's a File or some other data supported by FormData, it will not work, but then again, it is either that, or creating some kind of compound plain string key, `formData.set('address.line1', address[line1])`, something like that but dynamically generated strings, and then somehow reconstruct that on the server.

This is a FormData API requirement, not server actions though.
